src/descriptor/tr.rs

```diff
@@ -0,0 +1,201 @@
+// Tapscript
+
+// use super::{
+//     checksum::{desc_checksum, verify_checksum},
+//     DescriptorTrait,
+// };
+use bitcoin::hashes::_export::_core::fmt::Formatter;
+use expression::{self, FromTree, Tree};
+use std::sync::Arc;
+use std::{fmt, str::FromStr};
+use Segwitv0;
+use {miniscript::Miniscript, Error, MiniscriptKey};
+
+// TODO: Update this to infer version from descriptor.
+const VER: u8 = 0xc0;
+
+#[derive(Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub enum TapTree<Pk: MiniscriptKey> {
+    Tree(Arc<TapTree<Pk>>, Arc<TapTree<Pk>>),
+    Miniscript_(u8, Arc<Miniscript<Pk, Segwitv0>>),
+}
+
+#[derive(Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub struct Tr<Pk: MiniscriptKey> {
+    key_path: Pk,
```

Comment:
rename as `internal_key` and `tree`? `key_path` and `script_path` represent the ways in which we spend scripts, not the structures.